JAMES-3365 JMAP back reference resolution

chibenwa opened the pull request:
**Strategy**: before parsing ones method argument, rely on **play-json** JSON transformers to substitute back-references in the supplied arguments by the results of previously executed methods.

Advantage: 
 - this method is **fully** transparent to method developer, who just don't need to worry about back-reference resolving, else than aat argument parsing time, and recording result time. => Easy!
 - this method allows complex and arbitrary client supplied value. We don't need to guess one use cases

Disadantage:
 - No strong typing: one can for instance pass emailIds into mailboxIds

**Notes**: 
 - This work aims at demonstrating this approach works, and did not at all address error management (which might require heroic monad management skills).
 - I did not reuse existing JSON pointer lib given that JMAP syntax slightly differs from the spec.
 - I'm unsure the /xx/*/yy/*zz work as is...

Arsnael commented on server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/routes/ProcessingContext.scala:
```diff
@@ -80,46 +86,67 @@ class ProcessingContext {
+    val seq: Seq[JsResult[JsValue]] = array.map(a => backReferenceResolver(serializer).reads(a)).toSeq
+    seq.find(res => res.isError)
+      .getOrElse(JsSuccess(JsArray(seq.map(res => res.get))))
+  }
```
I think the indent was initially right?

Arsnael commented on server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/routes/ProcessingContext.scala:
```diff
@@ -80,46 +86,67 @@ class ProcessingContext {
+    }).toMap
+
+    firstError.getOrElse(JsSuccess(JsObject(transformedMap)))
+  }
```
indent here too

Arsnael commented on server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/routes/ProcessingContext.scala:
```diff
@@ -80,46 +86,67 @@ class ProcessingContext {
+    firstError.getOrElse(JsSuccess(JsObject(transformedMap)))
+  }
+
+  private def resolveBackReference(serializer: Serializer)(entry: (String, JsValue)): Either[JsError, (String, JsValue)] = {
```
idem

chibenwa commented:
Forced pushed as I did some significant changes:

 - Fixed some tiny mistakes preventing things to work...
 - JMAPApiRoutes does all the JSON translation and invocation recording work. Method developper just don't need to worry anymore about back-reference resolution (at all!)
 - I did a separated contracts and added a few tests regardings errors

I feel pretty happy about this work now, so if you agree with the overall strategy then I will write more tests and eventually undraft this!

chibenwa commented:
Forced pushed to solve a simple conflict.

I added some tests for path resolution (works great!)

chibenwa commented:
Deployment test failed

mvn tests :green_apple: 

test this please

chibenwa commented on server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/routes/ProcessingContext.scala:
```diff
@@ -19,18 +19,157 @@
+    case _ => JsError("Expecting an array")
+  }
+
+  private def expendArray(evaluationResults: List[JsResult[JsValue]]) = {
```
Yes I relied on IDE extraction, but did not added the types... Sorry.

chibenwa commented on server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/routes/ProcessingContext.scala:
```diff
@@ -19,18 +19,157 @@
+case class InvalidResultReferenceException(message: String) extends IllegalArgumentException
+
 class ProcessingContext {
  private val creationIds: mutable.Map[ClientId, ServerId] = mutable.Map()
```
I don't understand.

chibenwa commented on server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/routes/ProcessingContext.scala:
```diff
@@ -19,18 +19,157 @@
 
+  def recordInvocation(invocation: Invocation): Unit = invocations.put(invocation.methodCallId, invocation)
+
+  def resolveBackReferences(serializer: Serializer, invocation: Invocation): Either[InvalidResultReferenceException, Invocation] =
```
We only have one Serializer class. Do you mean renaming the "serializer" parameter to "deserializer"?

chibenwa commented on server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/routes/JMAPApiRoutes.scala:
```diff
@@ -114,13 +114,14 @@ class JMAPApiRoutes (val authenticator: Authenticator,
       requestObject
         .methodCalls
-        .map(invocation => this.processMethodWithMatchName(requestObject.using.toSet, invocation, mailboxSession, processingContext))
+        .map(invocation => process(capabilities, mailboxSession, processingContext, invocation))
```
Contributions welcomed ;-)

Arsnael commented on server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/BackReferenceContract.scala:
```diff
@@ -0,0 +1,487 @@
+           |       "properties": ["id"],
+           |       "#ids": {
+           |         "resultOf":"c42",
+           |         "name":"Mailbox/set",
```
```suggestion
           |         "name":"Mailbox/get",
```

chibenwa reviewed (commented):
Nice thanks, I was trying to write something similar...

chibenwa commented:
Ready for review <3

chibenwa commented:
test this please (deployment test)

Arsnael commented on server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/BackReferenceContract.scala:
```diff
@@ -0,0 +1,487 @@
+         |                    },
+         |                    {
+         |                        "id": "6"
+         |                    }
```
Wouldn't it be less heavy and acceptable to write each items in the list like this, as we only have id anyway: `{"id": "6"},` ?

Arsnael commented on server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/routes/JMAPApiRoutes.scala:
```diff
@@ -129,12 +130,25 @@ class JMAPApiRoutes (val authenticator: Authenticator,
     }
   }
 
+  private def process(capabilities: Set[CapabilityIdentifier], mailboxSession: MailboxSession, processingContext: ProcessingContext, invocation: Invocation) = {
```
No return type defined?

chibenwa replied:
This comment is outdated

nvduc91 commented on server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/routes/ProcessingContext.scala:
```diff
@@ -19,18 +19,155 @@
+  }
+
+  private def resolveBackReferences(serializer: Serializer, array: collection.IndexedSeq[JsValue]): JsResult[JsValue] = {
+    val seq: Seq[JsResult[JsValue]] = array.map(a => backReferenceResolver(serializer).reads(a)).toSeq
```
`a` could is had better name?

chibenwa commented:
test this please

chibenwa commented:
`RabbitMQEventBusTest$ConcurrentTest.concurrentDispatchKeyShouldDeliverAllEventsToListenersWithSingleEventBus » RetryExhausted`

unrelated

test this please

chibenwa commented:
test this please

deployment test - can be considered :green_apple:

Arsnael commented:
Merged
